File: Rust/P4.rs
fn main(){
    let mut largest: i64 = 0;
    for i in 100..1000 {
        for j in i..1000{ //optimization, reduces duplicate calculations
            let t = i*j;
            if is_palindrome(t) && t > largest {
                largest = t;
            }
        }
    }

    println!("{}", largest);
}

fn is_palindrome(mut n: i64) -> bool {
    let mut divisor: i64 = 1;
    while n / divisor >= 10 { divisor *= 10; }

    while n != 0 {
        let msd = n / divisor; //Most Signifigant Digit
        let lsd = n % 10; //Lease Signifigant Digit

        if msd != lsd {return false;}

        n = (n % divisor) / 10; //Strip LSD and MSD
        divisor /= 100; //We stripped 2 digits, so divisor loses factor of 100
    }
    true
}

File: C/P4.c
#include <stdio.h>

int is_palindrome(int n);

int main(){
    int i, j;
    int largest = 0;
    int t;
    for(i = 100; i < 1000; i++){
        for(j = i; j < 1000; j++){ //optimization, reduce duplicate calcuations
            t = i*j;
            if(is_palindrome(t) == 1 & t > largest){
                largest = t;
            }
        }
    }
    printf("%d\n", largest);

return 0;
}

int is_palindrome(int n){
    int divisor = 1;

    while(n/divisor >= 10) { divisor *= 10; } //set divisor to be appropriate

    while(n != 0){
        int msd = n / divisor; //Most Signifigant Digit
        int lsd = n % 10; //Least Signifigant Digit

        if (msd != lsd){
            return 0;
        }

        n = (n % divisor) / 10; //remove MSD and LSD
        divisor /= 100; //because we stripped 2 digits, we need to reduce by 100
    }    

return 1;
}
